src/main/java/uk/gov/hmcts/darts/common/datamanagement/api/DataManagementFacadeImpl.java

```diff
@@ -1,125 +1,173 @@
+        if (CollectionUtils.isEmpty(storedEodEntities)) {
+            log.error("Supplied list of EodEntities does not have any that are stored");
+            throw new FileNotDownloadedException("Supplied list of EodEntities does not have any that are stored");
+        }
+        try {
+            return getDataFromStorage(storedEodEntities);
+        } catch (FileNotDownloadedException fnde) {
+            log.error("Could not retrieve file from any storage for eodId: {}", eodEntities.get(0).getId());
+            throw fnde;
+        }
+    }
 
-            // now fallback as we have not found any way of processing the download
-            processFallback(downloadableExternalObjectDirectories);
+    private DownloadResponseMetaData retrieveFileFromStorage(DatastoreContainerType datastoreType, ExternalObjectDirectoryEntity eodEntity,
+                                                             BlobContainerDownloadable container) throws FileNotDownloadedException {
+        try {
+            return container.downloadBlobFromContainer(datastoreType, eodEntity);
+        } catch (UncheckedIOException | BlobStorageException e) {
+            throw new FileNotDownloadedException(eodEntity.getExternalLocation(), datastoreType.name(), "Error downloading blob", e);
         }
     }
 
-    private void processDownloadResponse(DatastoreContainerType type, ExternalObjectDirectoryEntity processing,
-                                         DownloadableExternalObjectDirectories downloadableExternalObjectDirectories,
-                                         BlobContainerDownloadable container) {
-        if (!downloadableExternalObjectDirectories.getResponse().isSuccessfulDownload()) {
-            boolean success = false;
-            try {
-                success = container.downloadBlobFromContainer(type,
-                                                              processing, downloadableExternalObjectDirectories.getResponse());
-            } catch (UncheckedIOException e) {
-                log.error("Error occurred working out whether to continue", e);
+    /**
+     * Loop through each storage type in order to see if it has a matched EodEntity, and if it does, try to download the file from there, if it fails,
+     * move to the next one, if they all fail then throw a FileNotDownloadedException.
+     */
+    private DownloadResponseMetaData getDataFromStorage(List<ExternalObjectDirectoryEntity> storedEodEntities) throws FileNotDownloadedException {
+        List<DatastoreContainerType> storageOrder = storageOrderHelper.getStorageOrder();
+        StringBuilder logBuilder = new StringBuilder("Starting to search for files with " + storedEodEntities.size() + " eodEntities\n");
+
+        for (DatastoreContainerType datastoreContainerType : storageOrder) {
+            logBuilder.append("checking container " + datastoreContainerType.name() + "\n");
```

Comment:
Is it ok to use mulltiline logs, would it still appear as a single log entry in appinsights/dynatrace?

Reply:
i guess it would handle it in the same way that it handles a stack trace?
![image](https://github.com/hmcts/darts-api/assets/131666106/e56ff46a-cfb5-45e4-aa2d-79a85d0179bd)